File: Apps/KS_PDF_Studio/ks_pdf_studio_sdk.py
"""
KS PDF Studio Python SDK
Official Python client library for KS PDF Studio API.

Author: Kalponic Studio
Version: 2.0.0
License: MIT
"""

import os
import json
import time
import requests
from typing import Dict, List, Optional, Union, Any, BinaryIO
from pathlib import Path
from urllib.parse import urljoin, urlparse
import mimetypes


class KSAPIError(Exception):
    """Base exception for KS API errors."""

    def __init__(self, message: str, status_code: int = None, response_data: Dict = None):
        super().__init__(message)
        self.status_code = status_code
        self.response_data = response_data or {}


class KSAuthenticationError(KSAPIError):
    """Authentication error."""
    pass


class KSValidationError(KSAPIError):
    """Validation error."""
    pass


class KSRateLimitError(KSAPIError):
    """Rate limit exceeded error."""
    pass


class KSClient:
    """
    KS PDF Studio API client.

    Args:
        api_key: Your API key
        base_url: API base URL (default: https://api.kspdfstudio.com/v1)
        timeout: Request timeout in seconds (default: 30)
        max_retries: Maximum number of retries for failed requests (default: 3)
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.kspdfstudio.com/v1",
        timeout: int = 30,
        max_retries: int = 3
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.max_retries = max_retries

        self.session = requests.Session()
        self.session.headers.update({
            'X-API-Key': self.api_key,
            'User-Agent': 'KS-PDF-Studio-Python-SDK/2.0.0'
        })

        # Initialize API modules
        self.pdf = PDFAPI(self)
        self.ai = AIAPI(self)
        self.documents = DocumentsAPI(self)
        self.licenses = LicensesAPI(self)
        self.analytics = AnalyticsAPI(self)
        self.files = FilesAPI(self)
        self.batch = BatchAPI(self)
        self.webhooks = WebhooksAPI(self)

    def _request(
        self,
        method: str,
        endpoint: str,
        data: Dict = None,
        files: Dict = None,
        params: Dict = None,
        json_data: Dict = None
    ) -> Dict:
        """Make API request with error handling and retries."""
        url = urljoin(self.base_url + '/', endpoint.lstrip('/'))

        for attempt in range(self.max_retries + 1):
            try:
                if json_data is not None:
                    response = self.session.request(
                        method=method,
                        url=url,
                        json=json_data,
                        params=params,
                        timeout=self.timeout
                    )
                elif files:
                    response = self.session.request(
                        method=method,
                        url=url,
                        data=data,
                        files=files,
                        params=params,
                        timeout=self.timeout
                    )
                else:
                    response = self.session.request(
                        method=method,
                        url=url,
                        data=data,
                        params=params,
                        timeout=self.timeout
                    )

                return self._handle_response(response)

            except requests.RequestException as e:
                if attempt == self.max_retries:
                    raise KSAPIError(f"Request failed after {self.max_retries + 1} attempts: {e}")
                time.sleep(2 ** attempt)  # Exponential backoff

    def _handle_response(self, response: requests.Response) -> Dict:
        """Handle API response and raise appropriate exceptions."""
        try:
            data = response.json()
        except ValueError:
            data = {}

        if response.status_code == 401:
            raise KSAuthenticationError("Authentication failed", response.status_code, data)
        elif response.status_code == 400:
            raise KSValidationError("Validation error", response.status_code, data)
        elif response.status_code == 429:
            raise KSRateLimitError("Rate limit exceeded", response.status_code, data)
        elif not response.ok:
            error_msg = data.get('error', f"API error: {response.status_code}")
            raise KSAPIError(error_msg, response.status_code, data)

        return data

    def health_check(self) -> Dict:
        """Check API health status."""
        return self._request('GET', '/health')

    def get_info(self) -> Dict:
        """Get API information."""
        return self._request('GET', '/info')


class PDFAPI:
    """PDF operations API."""

    def __init__(self, client: KSClient):
        self.client = client

    def generate(
        self,
        content: str,
        template: str = "professional",
        title: str = None,
        options: Dict = None
    ) -> Dict:
        """
        Generate PDF from content.

        Args:
            content: Document content (markdown or plain text)
            template: PDF template ("professional", "academic", "simple")
            title: Document title
            options: Additional PDF options

        Returns:
            Dict containing file_id, download_url, etc.
        """
        data = {
            'content': content,
            'template': template
        }

        if title:
            data['title'] = title
        if options:
            data['options'] = options

        return self.client._request('POST', '/pdf/generate', json_data=data)

    def extract(
        self,
        file_path: Union[str, Path, BinaryIO],
        method: str = "text",
        pages: str = "all"
    ) -> Dict:
        """
        Extract text from PDF.

        Args:
            file_path: Path to PDF file or file-like object
            method: Extraction method ("text" or "ocr")
            pages: Pages to extract ("all" or comma-separated numbers)

        Returns:
            Dict containing extracted text and metadata
        """
        files = {}
        data = {'method': method, 'pages': pages}

        if hasattr(file_path, 'read'):
            files['file'] = ('file.pdf', file_path, 'application/pdf')
        else:
            file_path = Path(file_path)
            files['file'] = (file_path.name, open(file_path, 'rb'), 'application/pdf')

        try:
            return self.client._request('POST', '/pdf/extract', files=files, data=data)
        finally:
            # Close file if we opened it
            if 'file' in files and hasattr(files['file'][1], 'close'):
                files['file'][1].close()

    def watermark(
        self,
        file_path: Union[str, Path, BinaryIO],
        watermark_text: str,
        watermark_type: str = "text",
        opacity: float = 0.3,
        position: str = "diagonal"
    ) -> Dict:
        """
        Add watermark to PDF.

        Args:
            file_path: Path to PDF file or file-like object
            watermark_text: Text to use as watermark
            watermark_type: "text" or "image"
            opacity: Opacity level (0-1)
            position: Watermark position

        Returns:
            Dict containing watermarked file info
        """
        files = {}
        data = {
            'watermark_text': watermark_text,
            'watermark_type': watermark_type,
            'opacity': opacity,
            'position': position
        }

        if hasattr(file_path, 'read'):
            files['file'] = ('file.pdf', file_path, 'application/pdf')
        else:
            file_path = Path(file_path)
            files['file'] = (file_path.name, open(file_path, 'rb'), 'application/pdf')

        try:
            return self.client._request('POST', '/pdf/watermark', files=files, data=data)
        finally:
            if 'file' in files and hasattr(files['file'][1], 'close'):
                files['file'][1].close()

    def merge(self, file_paths: List[Union[str, Path]], output_name: str = None) -> Dict:
        """
        Merge multiple PDFs.

        Args:
            file_paths: List of PDF file paths
            output_name: Name for merged file

        Returns:
            Dict containing merged file info
        """
        files = {}
        data = {}

        if output_name:
            data['output_name'] = output_name

        for i, file_path in enumerate(file_paths):
            file_path = Path(file_path)
            files[f'files'] = (file_path.name, open(file_path, 'rb'), 'application/pdf')

        try:
            return self.client._request('POST', '/pdf/merge', files=files, data=data)
        finally:
            for file_tuple in files.values():
                if hasattr(file_tuple[1], 'close'):
                    file_tuple[1].close()

    def split(
        self,
        file_path: Union[str, Path, BinaryIO],
        split_type: str = "pages",
        pages_per_file: int = None,
        ranges: List[str] = None
    ) -> Dict:
        """
        Split PDF into multiple files.

        Args:
            file_path: Path to PDF file or file-like object
            split_type: "pages" or "ranges"
            pages_per_file: Pages per file (for "pages" type)
            ranges: Page ranges like ["1-5", "6-10"] (for "ranges" type)

        Returns:
            Dict containing split file info
        """
        files = {}
        data = {'split_type': split_type}

        if hasattr(file_path, 'read'):
            files['file'] = ('file.pdf', file_path, 'application/pdf')
        else:
            file_path = Path(file_path)
            files['file'] = (file_path.name, open(file_path, 'rb'), 'application/pdf')

        if split_type == "pages" and pages_per_file:
            data['pages_per_file'] = pages_per_file
        elif split_type == "ranges" and ranges:
            data['ranges'] = json.dumps(ranges)

        try:
            return self.client._request('POST', '/pdf/split', files=files, data=data)
        finally:
            if 'file' in files and hasattr(files['file'][1], 'close'):
                files['file'][1].close()


class AIAPI:
    """AI operations API."""

    def __init__(self, client: KSClient):
        self.client = client

    def enhance(
        self,
        content: str,
        enhancement_type: str = "general",
        options: Dict = None
    ) -> Dict:
        """
        Enhance content with AI.

        Args:
            content: Content to enhance
            enhancement_type: Type of enhancement
            options: Additional options

        Returns:
            Dict containing enhanced content
        """
        data = {
            'content': content,
            'enhancement_type': enhancement_type
        }

        if options:
            data['options'] = options

        return self.client._request('POST', '/ai/enhance', json_data=data)

    def generate(
        self,
        prompt: str,
        content_type: str = "article",
        length: str = "medium",
        style: str = "professional",
        audience: str = "general",
        include_examples: bool = True
    ) -> Dict:
        """
        Generate content with AI.

        Args:
            prompt: Content generation prompt
            content_type: Type of content to generate
            length: Content length ("short", "medium", "long")
            style: Writing style
            audience: Target audience
            include_examples: Whether to include examples

        Returns:
            Dict containing generated content
        """
        data = {
            'prompt': prompt,
            'content_type': content_type,
            'length': length,
            'style': style,
            'audience': audience,
            'include_examples': include_examples
        }

        return self.client._request('POST', '/ai/generate', json_data=data)

    def summarize(
        self,
        content: str,
        summary_type: str = "key_points",
        length: str = "medium",
        format: str = "paragraph"
    ) -> Dict:
        """
        Summarize content with AI.

        Args:
            content: Content to summarize
            summary_type: Type of summary
            length: Summary length
            format: Output format

        Returns:
            Dict containing summary
        """
        data = {
            'content': content,
            'summary_type': summary_type,
            'length': length,
            'format': format
        }

        return self.client._request('POST', '/ai/summarize', json_data=data)


class DocumentsAPI:
    """Document management API."""

    def __init__(self, client: KSClient):
        self.client = client

    def list(
        self,
        limit: int = 50,
        offset: int = 0,
        sort_by: str = "created",
        sort_order: str = "desc"
    ) -> Dict:
        """List documents."""
        params = {
            'limit': limit,
            'offset': offset,
            'sort_by': sort_by,
            'sort_order': sort_order
        }
        return self.client._request('GET', '/documents', params=params)

    def create(self, title: str, content: str, template: str = "professional", metadata: Dict = None) -> Dict:
        """Create a new document."""
        data = {
            'title': title,
            'content': content,
            'template': template
        }
        if metadata:
            data['metadata'] = metadata

        return self.client._request('POST', '/documents', json_data=data)

    def get(self, document_id: str) -> Dict:
        """Get document details."""
        return self.client._request('GET', f'/documents/{document_id}')

    def update(self, document_id: str, **updates) -> Dict:
        """Update document."""
        return self.client._request('PUT', f'/documents/{document_id}', json_data=updates)

    def delete(self, document_id: str) -> Dict:
        """Delete document."""
        return self.client._request('DELETE', f'/documents/{document_id}')


class LicensesAPI:
    """License management API."""

    def __init__(self, client: KSClient):
        self.client = client

    def create(
        self,
        user_id: str,
        user_name: str,
        user_email: str,
        license_type: str,
        content_id: str,
        content_title: str,
        duration_days: int = None,
        max_uses: int = None,
        features: List[str] = None,
        watermark_text: str = None
    ) -> Dict:
        """Create a new license."""
        data = {
            'user_id': user_id,
            'user_name': user_name,
            'user_email': user_email,
            'license_type': license_type,
            'content_id': content_id,
            'content_title': content_title
        }

        if duration_days:
            data['duration_days'] = duration_days
        if max_uses:
            data['max_uses'] = max_uses
        if features:
            data['features'] = features
        if watermark_text:
            data['watermark_text'] = watermark_text

        return self.client._request('POST', '/licenses', json_data=data)

    def validate(self, license_key: str, content_id: str = None, user_id: str = None) -> Dict:
        """Validate a license key."""
        data = {'license_key': license_key}
        if content_id:
            data['content_id'] = content_id
        if user_id:
            data['user_id'] = user_id

        return self.client._request('POST', '/licenses/validate', json_data=data)

    def get(self, license_id: str) -> Dict:
        """Get license details."""
        return self.client._request('GET', f'/licenses/{license_id}')


class AnalyticsAPI:
    """Analytics API."""

    def __init__(self, client: KSClient):
        self.client = client

    def usage(
        self,
        days: int = 30,
        user_id: str = None,
        license_id: str = None,
        event_type: str = None
    ) -> Dict:
        """Get usage analytics."""
        params = {'days': days}
        if user_id:
            params['user_id'] = user_id
        if license_id:
            params['license_id'] = license_id
        if event_type:
            params['event_type'] = event_type

        return self.client._request('GET', '/analytics/usage', params=params)

    def revenue(
        self,
        days: int = 30,
        license_type: str = None,
        group_by: str = "day"
    ) -> Dict:
        """Get revenue analytics."""
        params = {'days': days, 'group_by': group_by}
        if license_type:
            params['license_type'] = license_type

        return self.client._request('GET', '/analytics/revenue', params=params)


class FilesAPI:
    """File operations API."""

    def __init__(self, client: KSClient):
        self.client = client

    def upload(self, file_path: Union[str, Path], category: str = "document") -> Dict:
        """Upload a file."""
        file_path = Path(file_path)
        with open(file_path, 'rb') as f:
            files = {'file': (file_path.name, f, mimetypes.guess_type(file_path.name)[0])}
            data = {'category': category}
            return self.client._request('POST', '/files/upload', files=files, data=data)

    def download(self, file_id: str, output_path: Union[str, Path]) -> None:
        """Download a file."""
        response = self.client.session.get(
            urljoin(self.client.base_url + '/', f'/files/download/{file_id}'),
            timeout=self.client.timeout
        )

        if response.status_code == 401:
            raise KSAuthenticationError("Authentication failed", response.status_code)
        elif not response.ok:
            raise KSAPIError(f"Download failed: {response.status_code}", response.status_code)

        output_path = Path(output_path)
        output_path.parent.mkdir(parents=True, exist_ok=True)

        with open(output_path, 'wb') as f:
            f.write(response.content)


class BatchAPI:
    """Batch operations API."""

    def __init__(self, client: KSClient):
        self.client = client

    def process(self, operations: List[Dict], options: Dict = None) -> Dict:
        """Process multiple operations in batch."""
        data = {'operations': operations}
        if options:
            data['options'] = options

        return self.client._request('POST', '/batch/process', json_data=data)

    def status(self, batch_id: str) -> Dict:
        """Get batch processing status."""
        return self.client._request('GET', f'/batch/status/{batch_id}')


class WebhooksAPI:
    """Webhooks API."""

    def __init__(self, client: KSClient):
        self.client = client

    def list(self) -> Dict:
        """List webhooks."""
        return self.client._request('GET', '/webhooks')

    def create(self, url: str, events: List[str], secret: str = None) -> Dict:
        """Create a webhook."""
        data = {
            'url': url,
            'events': events
        }
        if secret:
            data['secret'] = secret

        return self.client._request('POST', '/webhooks', json_data=data)

    def delete(self, webhook_id: str) -> Dict:
        """Delete a webhook."""
        return self.client._request('DELETE', f'/webhooks/{webhook_id}')


# Convenience functions
def create_client(api_key: str = None, **kwargs) -> KSClient:
    """
    Create KS PDF Studio API client.

    Args:
        api_key: API key (can also be set via KS_API_KEY environment variable)
        **kwargs: Additional client arguments

    Returns:
        KSClient instance
    """
    api_key = api_key or os.environ.get('KS_API_KEY')
    if not api_key:
        raise ValueError("API key is required. Set KS_API_KEY environment variable or pass api_key parameter.")

    return KSClient(api_key, **kwargs)


# Example usage
if __name__ == "__main__":
    # Example usage (requires valid API key)
    try:
        client = create_client("your_api_key_here")

        # Health check
        health = client.health_check()
        print(f"API Status: {health['status']}")

        # Generate PDF
        result = client.pdf.generate(
            content="# Hello World\n\nThis is a test document.",
            template="professional",
            title="Test Document"
        )
        print(f"Generated PDF: {result['file_id']}")

        # Download PDF
        client.files.download(result['file_id'], "output.pdf")
        print("PDF downloaded as output.pdf")

    except KSAPIError as e:
        print(f"API Error: {e}")
        if e.response_data:
            print(f"Details: {e.response_data}")

File: Apps/KS_PDF_Studio/ks_pdf_studio_sdk.js
/**
 * KS PDF Studio JavaScript SDK
 * Official JavaScript client library for KS PDF Studio API
 *
 * @author Kalponic Studio
 * @version 2.0.0
 * @license MIT
 */

class KSAPIError extends Error {
  constructor(message, statusCode, responseData) {
    super(message);
    this.name = 'KSAPIError';
    this.statusCode = statusCode;
    this.responseData = responseData;
  }
}

class KSAuthenticationError extends KSAPIError {
  constructor(message, statusCode, responseData) {
    super(message, statusCode, responseData);
    this.name = 'KSAuthenticationError';
  }
}

class KSValidationError extends KSAPIError {
  constructor(message, statusCode, responseData) {
    super(message, statusCode, responseData);
    this.name = 'KSValidationError';
  }
}

class KSRateLimitError extends KSAPIError {
  constructor(message, statusCode, responseData) {
    super(message, statusCode, responseData);
    this.name = 'KSRateLimitError';
  }
}

class KSClient {
  /**
   * Create KS PDF Studio API client
   *
   * @param {Object} config Client configuration
   * @param {string} config.apiKey - Your API key
   * @param {string} [config.baseURL='https://api.kspdfstudio.com/v1'] - API base URL
   * @param {number} [config.timeout=30000] - Request timeout in milliseconds
   * @param {number} [config.maxRetries=3] - Maximum number of retries
   */
  constructor(config) {
    if (!config || !config.apiKey) {
      throw new Error('API key is required');
    }

    this.apiKey = config.apiKey;
    this.baseURL = config.baseURL || 'https://api.kspdfstudio.com/v1';
    this.timeout = config.timeout || 30000;
    this.maxRetries = config.maxRetries || 3;

    // Initialize API modules
    this.pdf = new PDFAPI(this);
    this.ai = new AIAPI(this);
    this.documents = new DocumentsAPI(this);
    this.licenses = new LicensesAPI(this);
    this.analytics = new AnalyticsAPI(this);
    this.files = new FilesAPI(this);
    this.batch = new BatchAPI(this);
    this.webhooks = new WebhooksAPI(this);
  }

  /**
   * Make API request with error handling and retries
   * @private
   */
  async _request(method, endpoint, data, options = {}) {
    const url = `${this.baseURL}${endpoint}`;
    const requestTimeout = options.timeout || this.timeout;
    const retries = options.retries || this.maxRetries;

    const headers = {
      'X-API-Key': this.apiKey,
      'User-Agent': 'KS-PDF-Studio-JS-SDK/2.0.0'
    };

    let lastError;

    for (let attempt = 0; attempt <= retries; attempt++) {
      try {
        let response;

        if (data instanceof FormData) {
          response = await fetch(url, {
            method,
            headers: {
              'X-API-Key': this.apiKey,
              'User-Agent': 'KS-PDF-Studio-JS-SDK/2.0.0'
            },
            body: data,
            signal: AbortSignal.timeout(requestTimeout)
          });
        } else {
          headers['Content-Type'] = 'application/json';
          response = await fetch(url, {
            method,
            headers,
            body: data ? JSON.stringify(data) : undefined,
            signal: AbortSignal.timeout(requestTimeout)
          });
        }

        return await this._handleResponse(response);

      } catch (error) {
        lastError = error;

        if (attempt < retries) {
          // Exponential backoff
          await new Promise(resolve => setTimeout(resolve, Math.pow(2, attempt) * 1000));
        }
      }
    }

    throw new KSAPIError(`Request failed after ${retries + 1} attempts: ${lastError.message}`);
  }

  /**
   * Handle API response and raise appropriate exceptions
   * @private
   */
  async _handleResponse(response) {
    let data;

    try {
      const contentType = response.headers.get('content-type');
      if (contentType && contentType.includes('application/json')) {
        data = await response.json();
      } else {
        data = await response.text();
      }
    } catch {
      data = {};
    }

    if (response.status === 401) {
      throw new KSAuthenticationError('Authentication failed', response.status, data);
    } else if (response.status === 400) {
      throw new KSValidationError('Validation error', response.status, data);
    } else if (response.status === 429) {
      throw new KSRateLimitError('Rate limit exceeded', response.status, data);
    } else if (!response.ok) {
      const errorMsg = (data && data.error) || `API error: ${response.status}`;
      throw new KSAPIError(errorMsg, response.status, data);
    }

    return data;
  }

  /**
   * Check API health status
   */
  async healthCheck() {
    return this._request('GET', '/health');
  }

  /**
   * Get API information
   */
  async getInfo() {
    return this._request('GET', '/info');
  }
}

class PDFAPI {
  constructor(client) {
    this.client = client;
  }

  /**
   * Generate PDF from content
   * @param {Object} options
   * @param {string} options.content - Document content
   * @param {string} [options.template='professional'] - PDF template
   * @param {string} [options.title] - Document title
   * @param {Object} [options.options] - Additional options
   */
  async generate(options) {
    return this.client._request('POST', '/pdf/generate', options);
  }

  /**
   * Extract text from PDF
   * @param {File|Blob} file - PDF file
   * @param {Object} [options={}]
   * @param {string} [options.method='text'] - Extraction method
   * @param {string} [options.pages='all'] - Pages to extract
   */
  async extract(file, options = {}) {
    const formData = new FormData();
    formData.append('file', file);

    if (options.method) formData.append('method', options.method);
    if (options.pages) formData.append('pages', options.pages);

    return this.client._request('POST', '/pdf/extract', formData);
  }

  /**
   * Add watermark to PDF
   * @param {File|Blob} file - PDF file
   * @param {Object} options
   * @param {string} options.watermarkText - Watermark text
   * @param {string} [options.watermarkType='text'] - Watermark type
   * @param {number} [options.opacity=0.3] - Opacity level
   * @param {string} [options.position='diagonal'] - Watermark position
   */
  async watermark(file, options) {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('watermark_text', options.watermarkText);

    if (options.watermarkType) formData.append('watermark_type', options.watermarkType);
    if (options.opacity !== undefined) formData.append('opacity', options.opacity.toString());
    if (options.position) formData.append('position', options.position);

    return this.client._request('POST', '/pdf/watermark', formData);
  }

  /**
   * Merge multiple PDFs
   * @param {File[]|Blob[]} files - PDF files to merge
   * @param {string} [outputName] - Name for merged file
   */
  async merge(files, outputName) {
    const formData = new FormData();

    files.forEach((file, index) => {
      formData.append('files', file, `file${index + 1}.pdf`);
    });

    if (outputName) {
      formData.append('output_name', outputName);
    }

    return this.client._request('POST', '/pdf/merge', formData);
  }

  /**
   * Split PDF into multiple files
   * @param {File|Blob} file - PDF file
   * @param {Object} options
   * @param {string} [options.splitType='pages'] - Split type
   * @param {number} [options.pagesPerFile] - Pages per file
   * @param {string[]} [options.ranges] - Page ranges
   */
  async split(file, options) {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('split_type', options.splitType || 'pages');

    if (options.splitType === 'pages' && options.pagesPerFile) {
      formData.append('pages_per_file', options.pagesPerFile.toString());
    } else if (options.splitType === 'ranges' && options.ranges) {
      formData.append('ranges', JSON.stringify(options.ranges));
    }

    return this.client._request('POST', '/pdf/split', formData);
  }
}

class AIAPI {
  constructor(client) {
    this.client = client;
  }

  /**
   * Enhance content with AI
   * @param {Object} options
   * @param {string} options.content - Content to enhance
   * @param {string} [options.enhancementType='general'] - Enhancement type
   * @param {Object} [options.options] - Additional options
   */
  async enhance(options) {
    return this.client._request('POST', '/ai/enhance', options);
  }

  /**
   * Generate content with AI
   * @param {Object} options
   * @param {string} options.prompt - Generation prompt
   * @param {string} [options.contentType='article'] - Content type
   * @param {string} [options.length='medium'] - Content length
   * @param {string} [options.style='professional'] - Writing style
   * @param {string} [options.audience='general'] - Target audience
   * @param {boolean} [options.includeExamples=true] - Include examples
   */
  async generate(options) {
    return this.client._request('POST', '/ai/generate', options);
  }

  /**
   * Summarize content with AI
   * @param {Object} options
   * @param {string} options.content - Content to summarize
   * @param {string} [options.summaryType='key_points'] - Summary type
   * @param {string} [options.length='medium'] - Summary length
   * @param {string} [options.format='paragraph'] - Output format
   */
  async summarize(options) {
    return this.client._request('POST', '/ai/summarize', options);
  }
}

class DocumentsAPI {
  constructor(client) {
    this.client = client;
  }

  /**
   * List documents
   * @param {Object} [options={}]
   * @param {number} [options.limit=50] - Maximum documents to return
   * @param {number} [options.offset=0] - Number of documents to skip
   * @param {string} [options.sortBy='created'] - Sort field
   * @param {string} [options.sortOrder='desc'] - Sort order
   */
  async list(options = {}) {
    const params = new URLSearchParams();

    if (options.limit) params.append('limit', options.limit.toString());
    if (options.offset) params.append('offset', options.offset.toString());
    if (options.sortBy) params.append('sort_by', options.sortBy);
    if (options.sortOrder) params.append('sort_order', options.sortOrder);

    const endpoint = `/documents${params.toString() ? '?' + params.toString() : ''}`;
    return this.client._request('GET', endpoint);
  }

  /**
   * Create a new document
   * @param {Object} options
   * @param {string} options.title - Document title
   * @param {string} options.content - Document content
   * @param {string} [options.template='professional'] - Document template
   * @param {Object} [options.metadata] - Document metadata
   */
  async create(options) {
    return this.client._request('POST', '/documents', options);
  }

  /**
   * Get document details
   * @param {string} documentId - Document ID
   */
  async get(documentId) {
    return this.client._request('GET', `/documents/${documentId}`);
  }

  /**
   * Update document
   * @param {string} documentId - Document ID
   * @param {Object} updates - Document updates
   */
  async update(documentId, updates) {
    return this.client._request('PUT', `/documents/${documentId}`, updates);
  }

  /**
   * Delete document
   * @param {string} documentId - Document ID
   */
  async delete(documentId) {
    return this.client._request('DELETE', `/documents/${documentId}`);
  }
}

class LicensesAPI {
  constructor(client) {
    this.client = client;
  }

  /**
   * Create a new license
   * @param {Object} options - License options
   */
  async create(options) {
    return this.client._request('POST', '/licenses', options);
  }

  /**
   * Validate a license key
   * @param {Object} options
   * @param {string} options.licenseKey - License key to validate
   * @param {string} [options.contentId] - Content ID
   * @param {string} [options.userId] - User ID
   */
  async validate(options) {
    return this.client._request('POST', '/licenses/validate', options);
  }

  /**
   * Get license details
   * @param {string} licenseId - License ID
   */
  async get(licenseId) {
    return this.client._request('GET', `/licenses/${licenseId}`);
  }
}

class AnalyticsAPI {
  constructor(client) {
    this.client = client;
  }

  /**
   * Get usage analytics
   * @param {Object} [options={}]
   * @param {number} [options.days=30] - Number of days to analyze
   * @param {string} [options.userId] - Filter by user ID
   * @param {string} [options.licenseId] - Filter by license ID
   * @param {string} [options.eventType] - Filter by event type
   */
  async usage(options = {}) {
    const params = new URLSearchParams();

    if (options.days) params.append('days', options.days.toString());
    if (options.userId) params.append('user_id', options.userId);
    if (options.licenseId) params.append('license_id', options.licenseId);
    if (options.eventType) params.append('event_type', options.eventType);

    const endpoint = `/analytics/usage${params.toString() ? '?' + params.toString() : ''}`;
    return this.client._request('GET', endpoint);
  }

  /**
   * Get revenue analytics
   * @param {Object} [options={}]
   * @param {number} [options.days=30] - Number of days to analyze
   * @param {string} [options.licenseType] - Filter by license type
   * @param {string} [options.groupBy='day'] - Group results by
   */
  async revenue(options = {}) {
    const params = new URLSearchParams();

    if (options.days) params.append('days', options.days.toString());
    if (options.licenseType) params.append('license_type', options.licenseType);
    if (options.groupBy) params.append('group_by', options.groupBy);

    const endpoint = `/analytics/revenue${params.toString() ? '?' + params.toString() : ''}`;
    return this.client._request('GET', endpoint);
  }
}

class FilesAPI {
  constructor(client) {
    this.client = client;
  }

  /**
   * Upload a file
   * @param {File|Blob} file - File to upload
   * @param {Object} [options={}]
   * @param {string} [options.category='document'] - File category
   */
  async upload(file, options = {}) {
    const formData = new FormData();
    formData.append('file', file);

    if (options.category) {
      formData.append('category', options.category);
    }

    return this.client._request('POST', '/files/upload', formData);
  }

  /**
   * Download a file
   * @param {string} fileId - File ID
   * @returns {Promise<Blob>} File blob
   */
  async download(fileId) {
    const url = `${this.client.baseURL}/files/download/${fileId}`;

    const response = await fetch(url, {
      headers: {
        'X-API-Key': this.client.apiKey,
        'User-Agent': 'KS-PDF-Studio-JS-SDK/2.0.0'
      }
    });

    if (response.status === 401) {
      throw new KSAuthenticationError('Authentication failed', response.status);
    } else if (!response.ok) {
      throw new KSAPIError(`Download failed: ${response.status}`, response.status);
    }

    return response.blob();
  }

  /**
   * Download file as URL (for browser use)
   * @param {string} fileId - File ID
   * @returns {Promise<string>} Object URL
   */
  async downloadAsURL(fileId) {
    const blob = await this.download(fileId);
    return URL.createObjectURL(blob);
  }
}

class BatchAPI {
  constructor(client) {
    this.client = client;
  }

  /**
   * Process multiple operations in batch
   * @param {Object} options
   * @param {Array} options.operations - Operations to process
   * @param {Object} [options.options] - Batch options
   */
  async process(options) {
    return this.client._request('POST', '/batch/process', options);
  }

  /**
   * Get batch processing status
   * @param {string} batchId - Batch ID
   */
  async status(batchId) {
    return this.client._request('GET', `/batch/status/${batchId}`);
  }
}

class WebhooksAPI {
  constructor(client) {
    this.client = client;
  }

  /**
   * List webhooks
   */
  async list() {
    return this.client._request('GET', '/webhooks');
  }

  /**
   * Create a webhook
   * @param {Object} options
   * @param {string} options.url - Webhook URL
   * @param {string[]} options.events - Events to listen for
   * @param {string} [options.secret] - Webhook secret
   */
  async create(options) {
    return this.client._request('POST', '/webhooks', options);
  }

  /**
   * Delete a webhook
   * @param {string} webhookId - Webhook ID
   */
  async delete(webhookId) {
    return this.client._request('DELETE', `/webhooks/${webhookId}`);
  }
}

// Convenience function to create client
function createClient(config) {
  return new KSClient(config);
}

// Export for different environments
if (typeof module !== 'undefined' && module.exports) {
  // Node.js/CommonJS
  module.exports = {
    KSClient,
    createClient,
    KSAPIError,
    KSAuthenticationError,
    KSValidationError,
    KSRateLimitError
  };
} else if (typeof define === 'function' && define.amd) {
  // AMD
  define([], function() {
    return {
      KSClient,
      createClient,
      KSAPIError,
      KSAuthenticationError,
      KSValidationError,
      KSRateLimitError
    };
  });
} else if (typeof window !== 'undefined') {
  // Browser global
  window.KSPDFStudio = {
    KSClient,
    createClient,
    KSAPIError,
    KSAuthenticationError,
    KSValidationError,
    KSRateLimitError
  };
}